### PR TITLE
Fix validation of nested arrays to allow indirect nesting

### PR DESCRIPTION
**Pull request by wilhuff**
With this change indirectly nested arrays are allowed. We still disallow
directly nested arrays.

Fixes internal bug b/66253451.

Port of firebase/firebase-ios-sdk#377

### CODE REVIEW

**Comment by notsonotso** [comment]
This bug is holding us back. When will the fix hit the road?

**Comment by wilhuff** [comment]
Post merge it should be soon, but we can't comment on specific timelines, sorry.

File: packages/firestore/src/api/user_data_converter.ts [hunk 1]
```diff
@@ -154,36 +159,52 @@ class ParseContext {
...
-    } else {
-      childPath = this.path == null ? null : this.path.child(field);
-    }
+  childContextForField(field: string): ParseContext {
+    const childPath = this.path == null ? null : this.path.child(field);
     const context = new ParseContext(
       this.dataSource,
       this.methodName,
       childPath,
+      false,
```
Comment (mikelehen) [inline packages/firestore/src/api/user_data_converter.ts]: Can you comment the boolean literal?  

`/*arrayElement=*/ false`

File: packages/firestore/src/api/user_data_converter.ts [hunk 2]
```diff
@@ -154,36 +159,52 @@ class ParseContext {
...
+    return context;
+  }
+
+  childContextForArray(index: number): ParseContext {
+    // TODO(b/34871131): We don't support array paths right now; so make path
+    // null.
+    return new ParseContext(
+      this.dataSource,
+      this.methodName,
+      null,
```
Comment (mikelehen) [inline packages/firestore/src/api/user_data_converter.ts]: Comment literals?
```
/*path=*/ null,
/*arrayElement=*/ true
```

**Comment by mikelehen** [comment]
Josh, can you approve the .gitignore change?

**Comment by wilhuff** [comment]
Fixes https://github.com/firebase/firebase-js-sdk/issues/193

**Pull request merged**